src/core/ext/filters/client_channel/client_channel.cc

```diff
@@ -1792,12 +1827,28 @@ void ChannelData::OnResolverError(grpc_error* error) {
   if (lb_policy_ == nullptr) {
     grpc_error* state_error = GRPC_ERROR_CREATE_REFERENCING_FROM_STATIC_STRING(
         "Resolver transient failure", &error, 1);
+    {
+      MutexLock lock(&resolution_mu_);
+      // Update resolver transient failure.
+      GRPC_ERROR_UNREF(resolver_transient_failure_error_);
+      resolver_transient_failure_error_ = GRPC_ERROR_REF(state_error);
+      // Process calls that were queued waiting for the resolver result.
+      for (QueuedCall* call = resolver_queued_calls_; call != nullptr;
+           call = call->next) {
+        grpc_call_element* elem = call->elem;
+        CallData* calld = static_cast<CallData*>(elem->call_data);
+        grpc_error* error = GRPC_ERROR_NONE;
+        if (calld->CheckResolutionLocked(elem, &error)) {
+          calld->AsyncResolutionDone(elem, error);
+        }
+      }
+    }
+    // Update connectivity state.
     UpdateStateAndPickerLocked(
```

Comment:
Which lock is being held at this point, and do we have a proof of non-inversion wrt the other lock held just above it?

Reply:
As mentioned above, this isn't actually holding any lock; it's being invoked from within the `WorkSerializer`.  There's no inversion issue here, because a `WorkSerializer` is effectively reentrant.